src/modules/constructRelationalGraph.js

```diff
@@ -0,0 +1,252 @@
+// @flow
+import Doctor from '../resources/Doctor.json';
+import Patient from '../resources/Patient.json';
+import Department from '../resources/Department.json';
+
+import type {
+  JoinCondition,
+  SelectionCondition,
+  Expression,
+  Graph,
+} from './types';
+
+let canOptimize = true;
+
+// Create a node of the table name with relevant fields
+const addNode = (
+  graph: Graph,
+  tableName: string,
+  selection?: ?SelectionCondition
+) => {
+  if (!(tableName in graph)) {
+    graph[tableName] = {
+      selections: [],
+      edges: {},
+    };
+  }
+  if (selection) {
+    graph[tableName]['selections'].push(selection);
+  }
+  return tableName;
+};
+
+// Add an edge for two tables if there's a valid join relation between them
+const addEdge = (
+  graph: Graph,
+  leftTable: string,
+  rightTable: string,
+  joinCondition: JoinCondition,
+  joinType: string,
+  leftSelection?: ?SelectionCondition,
+  rightSelection?: ?SelectionCondition
+) => {
+  addNode(graph, leftTable, leftSelection);
+  addNode(graph, rightTable, rightSelection);
+  if (!(rightTable in graph[leftTable]['edges'])) {
+    graph[leftTable]['edges'][rightTable] = [];
+  }
+  if (!(leftTable in graph[rightTable]['edges'])) {
+    graph[rightTable]['edges'][leftTable] = [];
+  }
+  const edge = {condition: joinCondition, type: joinType};
+  graph[leftTable]['edges'][rightTable].push(edge);
+  graph[rightTable]['edges'][leftTable].push(edge);
+};
+
+const conditionIsOnValidAttributes = (
+  attribute: string,
+  tables: Array<string>
+) => {
+  const columns: Array<string> = [];
+  for (const table of tables) {
+    switch (table) {
+      case 'Doctor':
+        columns.push(...Doctor.columns);
+        break;
+      case 'Patient':
+        columns.push(...Patient.columns);
+        break;
+      case 'Department':
+        columns.push(...Department.columns);
+        break;
+      default:
+        // pass
+        break;
+    }
+  }
+  return columns.includes(attribute);
+};
```

Comment:
It would be really nice to avoid having more code that depends on the specific tables that are being used. Can this not any existing code for doing this step?